src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -202,39 +202,50 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     if (hasCertificateAndKey)
                     {
                         bool hasCertReference = false;
-                        try
+                        if (sslAuthenticationOptions.CertificateContext != null && sslAuthenticationOptions.CertificateContext.IntermediateCertificates != null)
                         {
-                            certHandle!.DangerousAddRef(ref hasCertReference);
-                            using (X509Certificate2 cert = new X509Certificate2(certHandle.DangerousGetHandle()))
+                            // We got chain already built. Use it instead of building one here.
+                            if (!Ssl.AddExtraChainCertificates(context, sslAuthenticationOptions.CertificateContext.IntermediateCertificates))
                             {
-                                X509Chain? chain = null;
-                                try
+                                throw CreateSslException(SR.net_ssl_use_cert_failed);
+                            }
+                        }
+                        else
```

Comment:
```suggestion
                        else if (sslAuthenticationOptions.CertificateContext != null)
```

This logic shouldn't redundantly run for self-signed certs, or certs issued directly from roots.

I'd change the plumbing so that there's always a SslStreamCertificateContext here and just delete this.

Reply:
Based on other comment responses, I'll change my recommendation to starting this code with `SslStreamCertificateContext ctx = sslAuthenticationOptions.CertificateContext ?? SslStreamCertificateContext.Create(new X509Certificate2(certHandle.DangerousGetHandle());`

That'll eliminate the duplicate chain-building logic, so they don't have to be kept in sync.

Reply:
I like that suggestion. I was actually wondering if SslStreamCertificateContext.Create() should create it's own copy of given certificate to be in charge of lifetime management. e.g. avoid issues when we grab reference and it is disposed or manipulated somewhere else.

Reply:
Cloning certificates has.. problems.  I strongly recommend not cloning it, and leaving it as a caller problem to not dispose the cert.  On .NET Core the cert can't be repopulated to a new value, only "what you saw" or "disposed".

Reply:
updated and unified with client certs.